### PR TITLE
Ports+LibGL+LibSoftGPU: Support Monkey Island 4

### PR DESCRIPTION
**Pull request by GMTA**
This PR is a collection of commits working towards supporting the ScummVM port running Monkey Island 4 / Escape from Monkey Island. Most of the game seems playable, but FMV is still not working and results in red screens. This will be fixed in a future PR.

Changes include, but are not limited to:
* New pixel buffers for `LibSoftGPU` that result in a significant performance boost by standardizing on `BGRA8888`
* New `libmad` and `libmpeg2` ports
* Lots of smaller `LibGL` API fixes

Screenshots of ScummVM running monkey4:

image
image

### CODE REVIEW

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 1]
```diff
@@ -1251,28 +1249,22 @@ void Device::set_light_model_params(const LightModelParameters& lighting_model)
...
 }
 
 NonnullRefPtr<Image> Device::create_image(unsigned width, unsigned height, unsigned depth, unsigned levels, unsigned layers)
 {
-    VERIFY(width > 0);
-    VERIFY(height > 0);
-    VERIFY(depth > 0);
-    VERIFY(levels > 0);
-    VERIFY(layers > 0);
+    VERIFY(width > 0 && height > 0 && depth > 0 && levels > 0 && layers > 0);
```
Comment (sunverwerth) [inline Userland/Libraries/LibSoftGPU/Device.cpp:1267]: Isn't it better to have these as separate VERIFYs? Then we would actually know which of the arguments was zero on failure.

**Comment by sunverwerth** [comment]
LGTM except one little thing that's up for discussion

File: Userland/Libraries/LibSoftGPU/Device.cpp [hunk 2]
```diff
@@ -1258,33 +1257,33 @@ void Device::set_light_model_params(const LightModelParameters& lighting_model)
...
-    if (x < 0 || y < 0 || x >= m_render_target->width() || y >= m_render_target->height())
+    if (x < 0 || y < 0 || x >= m_frame_buffer->rect().width() || y >= m_frame_buffer->rect().height())
         return 1.0f;
 
-    return m_depth_buffer->scanline(y)[x];
+    return m_frame_buffer->depth_buffer()->scanline(y)[x];
 }
 
-NonnullRefPtr<Image> Device::create_image(ImageFormat format, unsigned width, unsigned height, unsigned depth, unsigned levels, unsigned layers)
+NonnullRefPtr<Image> Device::create_image(unsigned width, unsigned height, unsigned depth, unsigned levels, unsigned layers)
```
Comment (sunverwerth) [inline Userland/Libraries/LibSoftGPU/Device.cpp:1265]: Seeing as this interface will also be used for other drivers we should not remove the format parameter but simply verify here that we get passed a supported format. In the future we should return a list of supported image formats via the DeviceInfo struct.

**Review by sunverwerth — CHANGES_REQUESTED** [review]

File: Userland/Libraries/LibSoftGPU/Device.cpp (hunk 1 above)
Comment (GMTA) [reply Userland/Libraries/LibSoftGPU/Device.cpp]: I originally thought VERIFY had some more behind the scenes macro magic going on, but it's a simple `assert` - so I reverted this change. Thanks!

File: Userland/Libraries/LibSoftGPU/Device.cpp (hunk 2 above)
Comment (GMTA) [reply Userland/Libraries/LibSoftGPU/Device.cpp]: Agreed, partially restored this so we always provide `BGRA8888`.

**Review by Quaker762 — APPROVED** [review]
This looks great! LGTM!

Love seeing some of my hacked together uni code (e.g `glFrustum`) slowly get removed too haha!

**Review by sunverwerth — APPROVED** [review]

File: Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+template<typename T>
+class Typed2DBuffer final : public RefCounted<Typed2DBuffer<T>> {
+public:
+    static ErrorOr<NonnullRefPtr<Typed2DBuffer>> try_create(Gfx::IntSize const& size)
+    {
+        auto buffer = TRY(Typed3DBuffer<T>::try_create(size.width(), size.height(), 1));
+        return adopt_ref(*new Typed2DBuffer(buffer));
+    }
+
+    void fill(T value, Gfx::IntRect const rect) { m_buffer->fill(value, rect.left(), rect.right(), rect.top(), rect.bottom(), 0, 0); }
```
Comment (linusg) [inline Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h:32]: Is `Gfx::IntRect const rect` supposed to be a const reference?

File: Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    ALWAYS_INLINE T* scanline(int y) { return m_buffer->buffer_pointer(0, y, 0); }
+    ALWAYS_INLINE T const* scanline(int y) const { return m_buffer->buffer_pointer(0, y, 0); }
+
+    void blit_from_bitmap(Gfx::Bitmap const& bitmap, Gfx::IntRect const target) requires IsSame<T, u32>
```
Comment (linusg) [inline Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h:36]: Is `Gfx::IntRect const target` supposed to be a const reference?

File: Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            auto* buffer_scanline = scanline(y);
+            auto const* bitmap_scanline = bitmap.scanline(source_y++);
+
+            int source_x = 0;
+            for (int x = target.left(); x <= target.right(); ++x)
+                buffer_scanline[x] = bitmap_scanline[source_x++];
+        }
+    }
+
+    void blit_to_bitmap(Gfx::Bitmap& bitmap, Gfx::IntRect const target) const requires IsSame<T, u32>
```
Comment (linusg) [inline Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h]: same here

File: Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h (hunk 3 above)
Comment (GMTA) [reply Userland/Libraries/LibSoftGPU/Buffer/Typed2DBuffer.h]: Made these `IntRect`s all const references :-) thanks!

**Review by linusg — APPROVED** [review]
:tada:

**Pull request merged**